inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/region_yolo_gpu_ref.cl

```diff
@@ -12,94 +12,95 @@
 
-#include "include/common.cl"
-#include "include/data_types.cl"
+#include "include/include_all.cl"
 
+#if INPUT0_LAYOUT_BFYX
 #define IW INPUT0_SIZES[0]
 #define IH INPUT0_SIZES[1]
 #define IC INPUT0_SIZES[2]
 #define IB INPUT0_SIZES[3]
+#elif INPUT0_LAYOUT_BYXF
+#define IC INPUT0_SIZES[0]
+#define IW INPUT0_SIZES[1]
+#define IH INPUT0_SIZES[2]
+#define IB INPUT0_SIZES[3]
+#endif
 
 inline UNIT_TYPE FUNC(logistic_activate)(UNIT_TYPE x) {
     return 1. / (1. + exp(-x));
 }
 
-inline int FUNC(entry_index)(int width, int height, int coords, int classes,
-                       int outputs, int batch, int location,
-                       int entry) {
-    int n = location / (width * height);
-    int loc = location % (width * height);
-    return batch * outputs + n * width * height * (coords + classes + 1) +
-        entry * width * height + loc;
-}
+inline int FUNC(output_index)(int batch, int region_num, int xy, int feature_offset) {
+    int region_offset = region_num * (COORDS + CLASSES + 1);
 
 #if DO_SOFTMAX
-inline void FUNC(softmax_generic)(const __global UNIT_TYPE* src_data, __global UNIT_TYPE* dst_data,
-                            int B, int C, int W, int H, int i)
-{
-    for (int b = 0; b < B; b++) {
-        UNIT_TYPE max = src_data[b*C*H*W + i];
-        for (int c = 0; c < C; c++) {
-            UNIT_TYPE val = src_data[b*C*H*W + c*H*W + i];
-            if (val > max) max = val;
-        }
-
-        UNIT_TYPE expSum = 0;
-        for (int c = 0; c < C; c++) {
-            dst_data[b*C*H*W + c*H*W + i] = exp(src_data[b*C*H*W + c*H*W + i] - max);
-            expSum += dst_data[b*C*H*W + c*H*W + i];
-        }
-
-        for (int c = 0; c < C; c++) {
-            dst_data[b*C*H*W + c*H*W + i] = dst_data[b*C*H*W + c*H*W + i] / expSum;
-        }
-    }
-}
+    return OUTPUT_GET_INDEX(batch, (feature_offset + region_offset) * INPUT0_SIZE_X * INPUT0_SIZE_Y + xy, 1, 1);
+#else
+    int x_index = xy % INPUT0_SIZE_X;
+    int y_index = (xy / INPUT0_SIZE_X) % (INPUT0_SIZE_Y);
+    return OUTPUT_GET_INDEX(batch, feature_offset + region_offset, y_index, x_index);
 #endif
+}
 
 KERNEL (region_yolo_ref)(const __global UNIT_TYPE* input, __global UNIT_TYPE* output)
 {
-    int x = get_global_id(0);
+    int xy = get_global_id(0);
+    int region_num = get_global_id(1);
+    int batch = get_global_id(2);
+    int x_index = xy % INPUT0_SIZE_X;
+    int y_index = (xy / INPUT0_SIZE_X) % (INPUT0_SIZE_Y);
 
-#if DO_SOFTMAX
-    #define ACTUAL_NUM (NUM)
-    #define CONF_CLASSES (1)
-#else
-    #define ACTUAL_NUM (MASK_SIZE)
-    #define CONF_CLASSES (CLASSES+1)
-#endif
-    #define INPUTS_COUNT (IH * IW * ACTUAL_NUM * (CLASSES + COORDS + 1))
+    /// [x, y, width, height, objectness score, class score]
+    /// x,y
+    int region_offset = region_num * (COORDS + CLASSES + 1);
+    int in_i = INPUT0_GET_INDEX(batch, 0 + region_offset, y_index, x_index);
+    int out_i = FUNC_CALL(output_index)(batch, region_num, xy, 0);
```

Comment:
Why don't you call with `(batch, region_num, xy, region_offset)`? Then, you don't need to calculate `region_offset` in output_index function again.

Reply:
4th arg in output_index function means offset in region. It's a relative offset.
'region_offset' is a absolute offset for input.